vendor/parser-extended/parser.mly

```diff
@@ -3332,30 +3363,53 @@ type_parameters:
       { ps }
 ;
 
-layout:
-  ident { check_layout $loc($1) $1 }
+layout_annotation: (* : layout_annotation *)
+  ident { let loc = make_loc $sloc in
+          mkloc (check_layout ~loc $1) loc }
+;
+
+layout_attr:
+  COLON
+  layout_annotation
+    { $2 }
+;
+
+%inline type_param_with_layout:
+  name=mkrhs(tyvar_name_or_underscore)
+  attrs=attributes
+  COLON
+  layout=layout_annotation
+  { let descr = Ptyp_var (name, Some layout) in
+    mktyp ~loc:$sloc ~attrs descr }
 ;
 
 parenthesized_type_parameter:
     type_parameter { $1 }
-  | type_variance type_variable COLON layout
-      { {$2 with ptyp_attributes = [$4]}, $1 }
+  | type_variance type_param_with_layout
+    { $2, $1 }
 ;
 
 type_parameter:
     type_variance type_variable attributes
       { {$2 with ptyp_attributes = $3}, $1 }
 ;
 
-type_variable:
+%inline type_variable:
   mktyp(
-    QUOTE tyvar = ident
-      { Ptyp_var tyvar }
+    QUOTE tyvar = mkrhs(ident)
+      { Ptyp_var (tyvar, None) }
   | UNDERSCORE
       { Ptyp_any }
   ) { $1 }
 ;
 
+%inline tyvar_name_or_underscore:
+    QUOTE ident
+      { $2 }
+  | UNDERSCORE
+      { "_" }
```

Comment:
This will need to be updated to avoid the check for underscores in `fmt_type_var`.

Reply:
This has been updated.